### DISCUSSION TITLE
Self-Hosting The App on a VPS Server

### DISCUSSION THREAD

**Discussion by drinkyourcoffee** [0perationPrivacy/VoIP#49] [Show and tell]
**NOTE:**
**If anyone is having problems for self hosting please don't make an Bug-Feature report. Everyone will be redirected to here again. THANK YOU**
.
Docker CLOSED issues reports logs:
https://github.com/0perationPrivacy/VoIP/issues/18
https://github.com/0perationPrivacy/VoIP/issues/54
.

After a lot of trial and error I was able to get this app working on a self-hosted VPS server running Ubuntu. The connections come into the server running Apache and get proxied to nodejs which runs on localhost using Apache's proxy and proxy_http modules (these need to be activated using a2enmod). Was able to get this functioning using SSL and after a lot of playing around with Apache virtualhost files. Here are the software/packages that were helpful in getting this functioning for two-way communication:

- apache2 with proxy and proxy_http enabled
- a2enmod (to enable the above)
- a2ensite (to add virtualhost to apache)
- mongodb (https://docs.mongodb.com/manual/tutorial/install-mongodb-on-ubuntu/)
- mongo and these mongo commands for clearing the database while troubleshooting:
   - db.users.remove({})
   - db.settings.remove({})
   - db.messages.remove({})
   - db.media.remove({})
- NodeJS
- npm (needed it to install dotenv module)
- ufw (or another firewall if using non-standard ports)
- Let's Encrypt (for free SSL certificates)

I wasn't sure what the mongodb connection string was supposed to be on a self-hosted server. This resource helped configure that correctly:
https://stackoverflow.com/questions/62260612/how-to-get-the-mongodb-connection-string

Had to modify app.js a bit to have it use https and know where to find the SSL certificate (there's probably a better way to do this) like this:
```
const https = require('https'),
        fs = require('fs');

const options = {
        key: fs.readFileSync("/etc/letsencrypt/live/domain.goes.here/privkey.pem"),
        cert: fs.readFileSync("/etc/letsencrypt/live/domain.goes.here/cert.pem")
};

const server = https.createServer(options, app);
```
Make headers more secure and harden app against automated attacks:
```
const frameguard = require('frameguard')
app.use(frameguard({ action: 'SAMEORIGIN' }))
app.disable('x-powered-by');
```
and at the end had to tell the app to listen on localhost:
`server.listen(8888, "127.0.0.1");`

For the apache virtualhosts file I was getting 500 errors and other errors until using this combination of proxy settings:
```
        ProxyRequests On
        SSLEngine On
        SSLProxyEngine On
        SSLProxyVerify none
        SSLProxyCheckPeerCN off
        SSLProxyCheckPeerName off
        SSLProxyCheckPeerExpire off
        ProxyPass / https://localhost:8888/ Keepalive=On
        ProxyPassReverse / https://localhost:8888/
        SSLCipherSuite ALL:!aNULL:RC4+RSA:+HIGH:+MEDIUM:+LOW:+EXP:+eNULL
        SSLCertificateFile "/etc/letsencrypt/live/domain.goes.here/cert.pem"
        SSLCertificateKeyFile "/etc/letsencrypt/live/domain.goes.here/privkey.pem"
        SSLCertificateChainFile "/etc/letsencrypt/live/domain.goes.here/chain.pem"
```

Hopefully this is helpful to anyone trying to get this running on their own server. I'm sure there are better ways to do this and I'm open to suggestions for improvements, but the main goal was to get this working on a self-hosted server while using an encrypted connection. Testing has shown that both SMS/MMS work in both directions. I'd be happy to answer any questions for anyone trying to self-host this. I set this up with Telnyx, so have no idea about the Twilio side of things.

**Comment by drinkyourcoffee** [comment]
Turns out more work needed to be done to get the reverse proxy to also allow websockets for faster updates (no longer have to refresh the app regularly to get new notifications. The following needed to be added to the Apache virtual host to get reverse proxying for websockets working after enabling the proxy_wstunnel module for Apache (a2enmod proxy_wstunnel):

`        <Location /socket.io>
                RewriteEngine on
                RewriteCond %{QUERY_STRING} transport=websocket    [NC]
                RewriteRule /(.*) wss://localhost:8888/socket.io/$1 [P,L]
                ProxyPass http://localhost:8888/socket.io retry=0 timeout=30
                ProxyPassReverse http://localhost:8888/socket.io
                RewriteCond ${HTTP:Upgrade} websocket [NC]
                RewriteCond ${HTTP:Connection} upgrade [NC]
                RewriteRule .* "wss:/localhost:8888/$1" [P,L]
        </Location>`

Incoming messages now appear almost instantly in Firefox Focus on GrapheneOS as well Chrome and Firefox for desktop. Vanadium doesn't seem to handle the incoming notifications on GrapheneOS well. It will update if exiting the current chat and going back in but not without a refresh if staying in the current chat.

**Comment by bioscopic** [comment]
Do you have steps available (documentation) to can provide so one can try to replicate what you did? Would like to try what you were able to accomplish on a bare metal server running Ubuntu. On what you were able to accomplish?

**Reply by drinkyourcoffee** [reply]
It works on a VPS server running Ubuntu so should work on a bare metal server. Unfortunately, the steps sort of depend on what you have set up already on your server. You would need to install a web server stack that at least includes Apache (Nginx would likely work as well, but I had success with Apache 2.4), MongoDB, and node/nodejs. If we assume you're running Ubuntu 20.04, you can use this guide for getting Apache installed:
https://www.digitalocean.com/community/tutorials/how-to-install-the-apache-web-server-on-ubuntu-20-04

For MongoDB:
https://www.digitalocean.com/community/tutorials/how-to-install-mongodb-on-ubuntu-20-04

For NodeJS:
https://www.digitalocean.com/community/tutorials/how-to-install-node-js-on-ubuntu-20-04

You may also be able to install these through Ubuntu's package management system, but may not get the most recent versions this way.

Once those are installed and configured you can try starting the application with:
`node start app.js`

You'll need to be in the directory where app.js is located (directly below "frontend ") to run that successfully. You can also install and run pm2:
https://pm2.keymetrics.io/docs/usage/quick-start/

With that installed, while in the directory where app.js is located, you could run:
`pm2 start app.js`

There is no official documentation on this process as it will be different for each environment, but hopefully this will help you get started. Reach out if you have any other questions as I know there are plenty of roadblocks to run into when getting this running on a VPS or bare metal server.

**Reply by paulmarker1** [reply]
This is what I need!  Heroku is not connected to github for me right now. I would prefer to run it on my own hardware anyways.

**Comment by fightforprivacy** [comment]
What would need to be done to get official documentation from start to finish assuming one had a VPS? 

This is something that should be heavily prioritized. If this was placed on the front page of the website to gain traction, it would get lots of upvotes, but instead it is buried, and no one knows about it until they submit a feature request.

**Reply by drinkyourcoffee** [reply]
I think the main issue with providing self-hosting documentation is that VPS services can vary widely. The documentation I provided above assumes one is running Ubuntu on a VPS although the steps should be quite similar for something like CentOS. I'd be happy to help if there are any questions about getting the app functioning on a VPS.

**Reply by paulmarker1** [reply]
I think for a lot of people the goal would be to self host at home. Most people would not care. Would be great to have it in a docker container then it wouldnt matter much at all.

**Reply by ragelfundionctrumnar** [reply]
I don't know if anyone is still looking at this thread, but the ending of the heroku free plan finally got me off my butt.  After looking through the various self-hosting and container threads here and created a very simple docker-compose stack which you can find at https://github.com/ragelfundionctrumnar/VoIP

It bundles 0perationPrivacy/VoIP, along with a local mongodb (though you can comment that out and use the connection string for mongodb.net) and caddy functioning as a reverse proxy.

I installed it on a $4/mo VPS in about 10 minutes.  Most of the "work" is installing docker and docker-compose.
